get real user name from username and email (for submission as real user with drmaa)

bernt-matthias opened the pull request:
The main part of this PR is that the drmaa job submission uses the username or the email that is stored in galaxy for submission as real user. This is in particular useful if the galaxy instance gets its users from LDAP. I do not know if this is already useful for users of other cluster software than drmaa. 

A secondary change is to preserve the PATH, LD_LIBRARY_PATH .. variables for the sudo calls for the external script since otherwise these variables are not kept in the environment (sudo resets them). I guess this needs to be discussed since sudo does this reset on purpose. I needed this, since otherwise in my installation python (which is not in a default path) is not found anymore. 

Furthermore, the external runner worked for me only if the --assign_all_groups parameter was used. Would be nice if somebody could verify if this is useful, since I did not understand this parameter.

Important note: I'm quite new to galaxy and this is my 1st "serious" PR. I guess this needs proper testing.

bernt-matthias committed:
- get real user name from username and email
- sudo calls with preserving PATH, LD_LIBRARY_PATH ...

bernt-matthias committed:
- added path check to external chown script

bernt-matthias committed:
- removed my paths

bernt-matthias commented:
Just added a small change to external_chown_script.py, that might inrease security (I thought it fits to the PR, since it affects the cluster access). 

Our admins here were concerned that with sudo rights on this script it is possible to change the owner and group of every file in the system. They suggested to make a check which allows changes only in subdirectories of the galaxy installation directory. I have made this configurable, i.e. one can configure a list of directories. To be sure the path given on the command line is made absolute, i.e., it should not be possible to place links inside the directory that point somewhere else. 

The idea would be that the admins can review the paths and then remove write access to the file when they add the sudo permissions.

mvdbeek commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -1743,7 +1743,14 @@ def _change_ownership( self, username, gid ):
-        cmd = [ '/usr/bin/sudo', '-E', external_chown_script, self.working_directory, username, str( gid ) ]
+
+        cmd = [ '/usr/bin/sudo', '-E', ]
+        for v in ["PATH", "LD_LIBRARY_PATH", "PKG_CONFIG_PATH"]:
```
This shouldn't be necessary, as 'sudo -E' does preserve the environment.
This is a small snippet that you can use to verify this:
```python
import os
import subprocess


print('without -E:')
subprocess.call(['sudo', './echo.sh'], shell=False)
print('with -E:')
subprocess.call(['sudo', '-E', './echo.sh'], shell=False)
```

Write a echo.sh script and echo some environment variable only available to the user.
You'll see that only with '-E' you get the environment variable echoed correctly.

bernt-matthias replied:
The security policy of our cluster has the env_reset option enabled (see man 5 sudoers) which in particular resets PATH. This seems to be the default on CentOS 6.8. In addition the LD_LIBRARY_PATH and PKG_CONFIG_PATH are reset. I'm not sure if PKG_CONFIG_PATH is actually needed - I just added everything that was removed. 

I think that the policy of sudo to reset the PATH and LIBRARY_PATH (etc) is actually a good idea. Just imagine a script with sudo rights that calls a binary B oder uses a library L. If you have the possibility to modify the PATH or LIBRARY_PATH before sudoing you have essentially the right to do everything. 

I think this modification is likely not of interest for many (maybe only me). It could also be that it is completely wrong to solve my problem like this. The question would be how to proceed here?

mvdbeek replied:
Hmm, could you ask your admins to make an exception for the external_chown_script.py ?
If I look at https://unix.stackexchange.com/a/13246 it appear you can specify the env vars to keep for a specific path. In a sense you're already circumventing this policy by injecting the the environment variables anyway ... . If not I guess we can again make the list of env vars configurable in the galaxy.ini file, I think.

bernt-matthias replied:
For the PATH we also thought that it is sufficient to hard code the path to the python binary in the script. But then there is still the problem with the python shared object file which is not found. Maybe one of you has an idea? Maybe I could build a special statically linked python binary just for these scripts. Could be a better idea than adding a potential security hole to galaxy.

mvdbeek replied:
Well, I don't think forcing users to build special version of the python interpreter just for this purpose is great. I kind of like injecting PATH (and LD_LIBRARY_PATH -urgs-, if necessary) instead of propagating all variables with `sudo -E`.

bernt-matthias replied:
Would it be an option to call 

`/path/to/python external_chown_script.py ... ` 

instead of 

`external_chown_script.py ... ` (using the shellbang) 

Then we could just use the python of galaxy's virtual environment if this path is accessible from the code. But this still leaves the problem of the library path (I never understood why the python binary but not the shared object is copied to the venv).

natefoo replied:
I'd prefer that this list be configurable, but see also my comment about not defaulting to using sudo at all.

mvdbeek commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -1770,10 +1777,14 @@ def reclaim_ownership( self ):
-                pass
+
+            # try to get user from email address/username
+            for cand in [job.user.email.split('@')[0], job.user.username]:
```
I'd prefer this to be configurable via a variable in galaxy.ini (sth. like `real_system_username`, with options 'user_email' and 'username'), close to the `external_chown_script` variable.

bernt-matthias replied:
Sounds possible. 

The question is if this is necessary, since testing both adds only a little computation (I guess its the same as the if-statement that would be necessary when we make this configurable). 

For me both possibilities would be fine. 

If we leave it as is I could imagine in case that both ways to determine a user are successful a test checking if they yield the same user would be a good idea. 

Otherwise: Can somebody point me to examples showing me how to access configuration variables?

mvdbeek replied:
I'm worried about corner cases like users who register an email address whose first part matches another users' username ... who then would get billed for computation time.
And explicit is better then implicit.

mvdbeek replied:
For adding options to galaxy.ini you'd write your option in the galaxy.ini.sample (with a bit of documentation) and then you reference it in the config object [here](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config.py#L349). This will then be available as `app.config.your_new_option` (in this class it's `self.app.config.your_new_option`)

bernt-matthias replied:
OK. Will do it this way.

mvdbeek commented on lib/galaxy/jobs/runners/drmaa.py:
```diff
@@ -316,7 +316,15 @@ def stop_job( self, job ):
-                subprocess.Popen( [ '/usr/bin/sudo', '-E', kill_script, str( ext_id ), str( self.userid ) ], shell=False )
+
+                command = [ '/usr/bin/sudo', '-E', ]
+                for v in ["PATH", "LD_LIBRARY_PATH", "PKG_CONFIG_PATH"]:
```
Again, I don't understand why this is necessary, 'sudo -E' should really preserve the environment.

mvdbeek commented on scripts/external_chown_script.py:
```diff
@@ -1,14 +1,34 @@
 #!/usr/bin/env python
 import os
+import os.path
```
This import is not required.

bernt-matthias replied:
OK. I removed it.

mvdbeek commented on scripts/external_chown_script.py:
```diff
@@ -1,14 +1,34 @@
+# if set to None every file can be modified by the script.
+# this can increase security in particular if write access to this
+# script is removed by the admin.
+# ALLOWED_PATHS = [ job_working_directory, new_file_path ]
```
Can you quote `job_working_directory` and `new_file_path` ?

bernt-matthias replied:
Done

mvdbeek reviewed (changes requested):
The core of the changes looks good, I would just like to confirm that specifically injecting PATH, LD_LIBRARY_PATH and PKG_CONFIG_PATH is really necessary.

bernt-matthias committed:
- removed uneccessary import
- added quotes to comment

bernt-matthias committed:
made determination of actual user configurable

bernt-matthias commented:
Just made it configurable via galaxi.ini

natefoo commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -1743,7 +1743,14 @@ def _change_ownership( self, username, gid ):
         external_chown_script = self.get_destination_configuration("external_chown_script", None)
-        cmd = [ '/usr/bin/sudo', '-E', external_chown_script, self.working_directory, username, str( gid ) ]
+
+        cmd = [ '/usr/bin/sudo', '-E', ]
```
Hardcoded sudo is actually a remnant of the original implementation. Ideally, if one needs sudo it should just be part of the value of the corresponding `*_script` call, at which point you'd have full control over the environment variables passed or not passed.

bernt-matthias replied:
I dont get what you are suggesting. We could have a little bash helper/wrapper: 

dosudo.sh:

#!/bin/bash
sudo -E AND_WHATEVER_IS_NEEDED "$@" 

From python we could then call: dosudo.sh externalhelper.py ...

bernt-matthias replied:
I think I got it. I moved the sudo calls to galaxy.ini

natefoo commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -1770,10 +1777,21 @@ def reclaim_ownership( self ):
-                self.__user_system_pwent = pwd.getpwnam( job.user.email.split('@')[0] )
-            except:
-                pass
+
```
The extra whitespace here and below isn't needed (sorry for the nitpick).

bernt-matthias replied:
With the helper/wrapper it would be configurable (at a single place)

natefoo commented on lib/galaxy/tools/actions/upload_common.py:
```diff
@@ -282,8 +282,30 @@ def create_paramfile( trans, uploaded_datasets ):
         try:
-            pwent = pwd.getpwnam( trans.user.email.split('@')[0] )
-            cmd = [ '/usr/bin/sudo', '-E', trans.app.config.external_chown_script, path, pwent[0], str( pwent[3] ) ]
+            # get username from email/username
```
There is a lot of duplication here that should be moved to a single function or method called by all who need it.

bernt-matthias replied:
Indeed. Would the helper script solve this problem already? Otherwise could you be more precise what should be packed in a function (just getting the user or also calling the external scripts). Where would such a function fit? I.e. in which python module?

bernt-matthias committed:
moved sudo calls to galaxy.ini
removed spaces

bernt-matthias committed:
imroved external_chown_script now uses Popen and checks return code

bernt-matthias commented:
Currently remaining is the request of @natefoo to put the code getting the user in a function. The question is where to put this in the galaxy libraries?

Of course I should also extend the documentation.

bernt-matthias commented:
Can anybody tell me how to continue from here? I would like to finish this PR.

rhpvorderman committed:
Fix linting issues

rhpvorderman committed:
fix typo

rhpvorderman committed:
Add system_user_pwent function

rhpvorderman committed:
Add system_user_pwent function to upload and jobs

rhpvorderman commented:
Hi there,
We would also like to use this function so I have fixed the duplication of code by putting a new function in `model/__init__.py`
EDIT: Also removed the duplaction of code in the external_chown_script.py

Changes can be found [here](https://github.com/galaxyproject/galaxy/compare/dev...rhpvorderman:fix_duplication)

@bernt-matthias 
You can merge these changes into your branch with the following commands:
```Bash
# Go to your galaxy repo
git remote add rhpvorderman https://github.com/rhpvorderman/galaxy.git
git fetch rhpvorderman
git merge rhpvorderman/fix_duplication
# commit the changes and push them.
```
Thanks for the great code! This was exactly what we needed.

rhpvorderman committed:
Change paramter typo

rhpvorderman committed:
Fix removed duplication in external chown script

rhpvorderman committed:
Fix linting

bernt-matthias commented:
Thanks a lot to @rhpvorderman for the contributions.

mvdbeek commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -1741,9 +1741,9 @@ def __link_file_check( self ):
-        # FIXME: hardcoded path
         external_chown_script = self.get_destination_configuration("external_chown_script", None)
-        cmd = [ '/usr/bin/sudo', '-E', external_chown_script, self.working_directory, username, str( gid ) ]
+        cmd = external_chown_script.split()
```
`shlex.split(external_chown_script)` would be better here.

mvdbeek replied:
And you also need to check that `external_chown_script` is not `None`.

mvdbeek commented on lib/galaxy/tools/actions/upload_common.py:
```diff
@@ -282,8 +281,10 @@ def create_paramfile( trans, uploaded_datasets ):
-            cmd = [ '/usr/bin/sudo', '-E', trans.app.config.external_chown_script, path, pwent[0], str( pwent[3] ) ]
+            # get username from email/username
+            pwent = trans.user.system_user_pwent(trans.app.config.real_system_username)
+            cmd = trans.app.config.external_chown_script.split()
```
same here, `shlex.split(trans.app.config.external_chown_script)` would be preferable.

mvdbeek commented on lib/galaxy/model/__init__.py:
```diff
@@ -13,6 +13,7 @@
 import os
 import socket
 import time
+import pwd
```
this needs to go before `import time` to satisfy the linter.

rhpvorderman replied:
My bad! I think it should go before `import socket` to satisfy the alphabetic order.

mvdbeek commented on scripts/external_chown_script.py:
```diff
@@ -2,23 +2,49 @@
+        (stdoutdata, stderrdata) = p.communicate()
+        exitcode = p.returncode
+        if exitcode != 0:
+            sys.stderr.write("external_chown_script: could not chown\ncmd was %s\n" % " ".join(cmd))
```
you can simplify this to `sys.exit("external_chown_script: could not chown\ncmd was %s\n" % " ".join(cmd)")`.

rhpvorderman commented:
Thanks for the review @mvdbeek! If you are too busy @bernt-matthias or need help, please let me know and then I will pick it up We are eager to get galaxy running on our cluster and you have saved me lots of time in writing this.

bernt-matthias commented:
I can incorporate the changes.

bernt-matthias committed:
implemented the changes requested by @mvdbeek:

- check for external_chown_script == None
- switched from string split to shlex.split
- changed import order for pwd
- stderr + exit -> just exit

bernt-matthias commented on lib/galaxy/tools/actions/upload_common.py:
```diff
@@ -283,7 +284,7 @@ def _chown( path ):
         try:
             # get username from email/username
             pwent = trans.user.system_user_pwent(trans.app.config.real_system_username)
```
I was wondering if its possible to use the get_destination_configuration function here. If yes: what would be the access path. Otherwise how can we find out that the variable is not defined in the config?

rhpvorderman replied:
This is already done in the system_user_pwent function by the following code:
```Python
# Checks if real_system_username is either user_email or username
       else:
           log.warning("invalid configuration of real_system_username")
           system_user_pwent = None
```
A log warning should be issued when the variable is not defined.

bernt-matthias committed:
bugfix missing colon

mvdbeek commented on lib/galaxy/jobs/runners/drmaa.py:
```diff
@@ -315,8 +315,9 @@ def stop_job( self, job ):
             else:
-                # FIXME: hardcoded path
-                subprocess.Popen( [ '/usr/bin/sudo', '-E', kill_script, str( ext_id ), str( self.userid ) ], shell=False )
+                command = kill_script.split()
```
Here it would also be better to use `shlex.split(kill_script)`, I should have mentioned it in the previous review round.

mvdbeek commented on lib/galaxy/jobs/runners/drmaa.py:
```diff
@@ -357,16 +358,12 @@ def store_jobtemplate(self, job_wrapper, jt):
-        for script_argument in script_parts[1:]:
-            command.append(script_argument)
-
+        command = external_runjob_script.split()
```
shlex.split

bernt-matthias committed:
shlex.split also for external kill and run scripts

bernt-matthias committed:
added missing empty line

mvdbeek commented:
@galaxybot test this

mvdbeek commented on lib/galaxy/jobs/runners/drmaa.py:
```diff
@@ -357,16 +359,12 @@ def store_jobtemplate(self, job_wrapper, jt):
-
+        command = shlex.split( external_runjob_script )
         command.extend( [ str(username), jobtemplate_filename ] )
+        command.append( "--assign_all_groups"  )
```
You can add this within the galaxy.ini (`drmaa_external_runjob_script = sudo -E scripts/drmaa_external_runner.py --assign_all_groups`), so I would prefer to not have this hardcoded here.

nsoranzo commented on scripts/external_chown_script.py:
```diff
@@ -2,23 +2,48 @@
 
-def validate_paramters():
+
+def validate_paramaters():
```
Typo: `s/paramaters/parameters/`

nsoranzo commented on scripts/external_chown_script.py:
```diff
@@ -2,23 +2,48 @@
-    os.system('chgrp -Rh %s %s' % (gid, path))
+    path, galaxy_user_name, gid = validate_paramaters()
+    for cmd in [[ 'chown', '-Rh', galaxy_user_name, path ], [ 'chgrp', '-Rh', gid, path ]]:
+        p = os.subprocess.Popen(cmd, shell=False, stdout=os.subprocess.PIPE, stderr=os.subprocess.PIPE)
```
You should change `os.subprocess` to `subprocess` (3 times).

Commit:
bug fixes

- system_user_pwent: undefined return value when KeyError
- external_chown_script:
  * typo
  * subprocess instead of os.subprocess

Commit:
moved assign_all_groups parameter to config

bernt-matthias commented:
I will test this tomorrow on our galaxy instance. But I can only test the behavior for the case where the username is taken directly (not from the mail). I will report possible errors / success. Can somebody test the other case?

bernt-matthias committed:
Merge remote-tracking branch 'upstream/dev' into dev

bernt-matthias commented:
Unfortunately I could not test today because I did not have write access to the external scripts (I hope our admins change this by tomorrow). Therefore I started to document the changes in galaxy-hub. See https://github.com/galaxyproject/galaxy-hub/pull/270 . Maybe you could have a look -- I might have forgotten something.

nsoranzo commented on config/galaxy.ini.sample:
```diff
@@ -1306,9 +1306,17 @@ use_interactive = True
+#drmaa_external_killjob_script = sudo -E scripts/drmaa_external_killer.py
+#external_chown_script = sudo -E scripts/external_chown_script.py
+
+# For the job submission as the actual system user galaxy can extract 
```
`user galaxy` -> `user, Galaxy`

nsoranzo commented on config/galaxy.ini.sample:
```diff
@@ -1306,9 +1306,17 @@ use_interactive = True
+
+# For the job submission as the actual system user galaxy can extract 
+# the user name from the email address (actually the local-part before the @) 
+# or the usename which are both stored in the galaxy data base. 
```
`usename` -> `username` and `galaxy data base` -> `Galaxy database`

bernt-matthias committed:
fixed typos in galaxy.ini

nsoranzo commented:
Looks good to me, but I'm not a user of this feature. May be of interest for @moskalenko.

bernt-matthias commented:
I have successfully tested the code for univa gridengine using the username configuration. I cannot test it for the email configuration or other drmaa systems.

bernt-matthias committed:
Merge remote-tracking branch 'upstream/dev' into dev
